### PR TITLE
usb: stm32: added usb_dc_detach code

### PR DESCRIPTION
**Pull request by StefJar**
Now detaching means, that the transfer endpoints are closed.
The usb device is stoped and deinit.

The usb periphal wil go to deepsleep after. That will save
some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

### CODE REVIEW

**Comment by StefJar** [comment]
pls review
@erwango

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -846,8 +846,13 @@ int usb_dc_ep_mps(const u8_t ep)
 
 int usb_dc_detach(void)
 {
-	LOG_ERR("Not implemented");
-
+	// stop usb dev
+	HAL_PCD_Stop(&usb_dc_stm32_state.pcd);
+	// take care that the endpoints are closed (usb_dc_ep_disable)
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_IN);
+	HAL_PCD_EP_Close(&usb_dc_stm32_state.pcd, EP0_OUT);
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c]: is it strictly necessary? what about the other endpoints?

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -846,8 +846,13 @@ int usb_dc_ep_mps(const u8_t ep)
...
+	// now we deinit
+	HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
 	return 0;
```
Comment (loicpoulain) [inline drivers/usb/device/usb_dc_stm32.c]: maybe we should create a usb_dc_stm32_deinit that undo everything done in usb_dc_stm32_init. and call it form here instead.

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (StefJar) [reply drivers/usb/device/usb_dc_stm32.c]: As far as I have seen there are only EP0_IN and EP0_OUT in that code. Guess stopping the usb, closing the endpoints and deinit is the most explicit and clean ending.

File: drivers/usb/device/usb_dc_stm32.c (hunk 2 above)
Comment (StefJar) [reply drivers/usb/device/usb_dc_stm32.c]: I am not sure about that. I am not a expert on the STM HAL and its different implementations. So you guys know best and should decide.

**Review by ydamigos — COMMENTED** [review]
@StefJar How did you test the change? Which sample?
In PR #16026, when HAL_PCD_DeInit() was called, the application got stuck. Did you face a similar problem?
@jli157 FYI.

File: drivers/usb/device/usb_dc_stm32.c [hunk 3]
```diff
@@ -846,12 +846,9 @@ int usb_dc_ep_mps(const u8_t ep)
 
 int usb_dc_detach(void)
 {
-	// stop usb dev
 	HAL_PCD_Stop(&usb_dc_stm32_state.pcd);
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: We don't need do this as `HAL_PCD_DeInit()` calls `HAL_PDC_Stop()`

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_stm32.c (hunk 3 above)
Comment (StefJar) [reply drivers/usb/device/usb_dc_stm32.c]: so let's break it down to only the `HAL_PCD_DeInit` call?
What is with the endpoints?

Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm32.c]: > What is with the endpoints?

Probably application needs to close all the endpoints that it opened. As far as I understand, it depends when `usb_disable()` is called. 

I am searching out STM32 documentation for the correct sequence to disable USB device. So far no luck. @erwango Could you help on this?

Comment (StefJar) [reply drivers/usb/device/usb_dc_stm32.c]: > @StefJar How did you test the change? Which sample?
> In PR #16026, when HAL_PCD_DeInit() was called, the application got stuck. Did you face a similar problem?
> @jli157 FYI.

nope
it worked for me fine(on a stm32f412)

Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm32.c]: > it worked for me fine(on a stm32f412)

How did you test it? Which sample or test case?

Comment (StefJar) [reply drivers/usb/device/usb_dc_stm32.c]: I have a sequence to put the device into sleep mode/reset. Inside that sequence I am deinit the usb. This works like follows:
- call `usb_disable`
- unint pins via `stm32_setup_pins`
- unclock usb via `LL_APB2_GRP1_DisableClock(LL_AHB2_GRP1_PERIPH_OTGFS)`

as you can see it is more less handcrafted. The test was run on 5 units (stm32f412) and went well

**Comment by StefJar** [comment]
will put everything into one commit and resubmit it

**Comment by StefJar** [comment]
the current pull can be found here
https://github.com/zephyrproject-rtos/zephyr/pull/22767

**Pull request closed without merging**